DIPs/13NNN-WGB.md

```diff
@@ -0,0 +1,893 @@
+```
+S func(return S s)
+{
+    S s2;
+    return s2;  // error, can't return local by Move Ref
+}
+```
+
+### Returning an EMO by `ref`
+
+Returning a local EMO by `ref` is an error, just as for non-EMO objects.
+
+```
+ref S f(S s)
+{
+    return s; // error
+}
+
+ref S g()
+{
+    S s;
+    return s; // error
+}
+```
+
+### Passing an EMO by `ref`
+
+The semantics of passing an EMO by `ref` are the same as for non-EMO objects.
+For example,
+
+```
+void func(ref S);
+...
+S s;
+void func(s);
+```
+is allowed and the responsiblity of destructing `s` remains with the caller.
+
+
+### Assigning a `ref` EMO to a non-`ref` EMO
+
+Consider:
+```
+void func(ref S s)
+{
+    S t;
+    t = s;  // `s` is copied, not moved, to `t`
+}
+```
+This is because explicit `ref` to EMOs do not own the object they refer
+to, so the only choice is to copy it.
+
+The same semantics apply to passing a `ref` EMO to a non-`ref` parameter:
+```
+void func(S);
+
+void g(ref S s)
+{
+    func(s); // a copy of `s` is made
+}
+```
+
+### Forwarding
+
+Referring back to our desired forwarding function `fwd`:
+
+```
+ref S fwd(return ref S s) { return s; }
+```
+which we would like to be used to forward its argument like so:
+
+```
+void f(S s);
+...
+S s;
+f(fwd(s));
+f(fwd(S());
+```
+Ownership of the argument to fwd() is retained by the caller,
+and so the caller will be responsible for its destruction.
+When the call is made to f(), a copy is made.
+
+If `fwd()` is inlineable, however, the compiler is allowed to examine
+its implementation. If implementation actually does return only a ref to the
+parameter `s`, the copy can be elided if it was the last use of `s`.
+Much like NRVO, it is implementation
+dependent whether this copy elision takes place.
+
+
+
+### EMO and Garbage Collection
+
+The current semantics are defined so that a naive compacting garbage collector
+can move objects with a mere memcpy. The restriction is that no fields of an object
+can contain a pointer to that same object. Currently, there isn't a naive compacting
+collector for D, so this hasn't been a problem. A collector that tracks where the
+pointers are in an object would not have this issue.
+
+Ironically, because move constructors execute arbitrary code, this likely would interfere
+with the collector keeping track of the state of memory during a collection cycle,
+meaning that objects with move constructors would likely be pinned (treated as immoveable).
+
+
+### Class Objects
+
+Class objects cannot have move constructors or move assignment operators.
+
+
+### Last Use
+
+The *Last Use* of an lvalue is the last time in a path of execution that any read
+or write is done to the memory region referred to directly by the lvalue. The last use
+of variable declarations is identified by doing dataflow analysis on function local
+variables and function parameters that are passed by value. Global variables and `ref`
+parameters are not checked for last use. For example, for a given function `f`:
```

Comment:
How accurate is this going to be? 


```D
struct S {
   // emo (nice name btw)
}

struct T {
S* s;
extern(C++) S* getProp() { return s; } // what if this isn't visible to do detection
}

void f() {
    S s;
    S* sp;
    {
        auto t = T(&s);
        sp = t.getProp();
    }

    scope(exit) {
         fun(*sp);
    }

    fun(s); // is this considered last use by the compiler (even though it actually isn't)?
}
```

This would need a lot of code analysis to determine when last use _actually_ is.

_Disclaimer: I don't agree with this feature. But if you are going to put a turd on the table, it might as well be a shiny turd 💩._

Reply:
This situation is explicitly addressed in the DIP. You're taking the address of `s`.

Reply:
It is not addressed in the DIP. Your taking an address, within a nested scope. That scope ends, the address it took is escaped. There's no way for data flow analysis to be able to catch unless it is able to see the getProp() code. 

Go ahead and tell me how it should then, the actual implementation by the compiler. You probably won't be able to as it seems you don't understand what the DIP is actually suggestioning to do. You just have that some abstract idea of it, which is why you think this situation is covered simply be cause there's a section about pointers.

Reply:
There are ways around that as well. The amount of rules this is introducing is just crazy. And Walter was trying to use simplifying the language to remove some features. I understand why now, cause he plans to introduce all of this rats nest of complexity. What C++ has implemented is far better than this, it is much simpler and the user can actually reason about the code, and move the object when they want to. Not under some very limited and narrow circumstances, and only the last use at that.